packages/core/test/app/utils/merkle-tree.spec.ts

```diff
@@ -456,4 +472,180 @@ describe('OptimizedSparseMerkleTree', () => {
+        ZERO,
+        zeroData
+      )
+
+      assert(proof.value.equals(zeroData))
+      assert(proof.key.equals(ZERO))
+
+      assert(proof.siblings.length === 2)
+      let hash: Buffer = hashFunction(oneData)
+      assert(proof.siblings[1].equals(hash))
+      hash = hashFunction(hashBuffer.fill(zeroHash, 0, 32).fill(zeroHash, 32))
+      assert(proof.siblings[0].equals(hash))
+
+      assert(proof.rootHash.equals(await tree.getRootHash()))
+
+      // Check Proof for ONE
+      proof = await tree.getMerkleProof(ONE, oneData)
+
+      assert(proof.value.equals(oneData))
+      assert(proof.key.equals(ONE))
+
+      assert(proof.siblings.length === 2)
+      hash = hashFunction(zeroData)
+      assert(proof.siblings[1].equals(hash))
+      hash = hashFunction(hashBuffer.fill(zeroHash, 0, 32).fill(zeroHash, 32))
+      assert(proof.siblings[0].equals(hash))
+
+      assert(proof.rootHash.equals(await tree.getRootHash()))
+    })
+
+    it('gets merkle proof for non-empty siblings 0 & 2', async () => {
+      const tree: SparseMerkleTree = await createAndVerifyEmptyTreeDepthWithDepth(
+        db,
+        ZERO,
+        3
+      )
+
+      const zeroData: Buffer = Buffer.from('ZERO 0')
+      await tree.update(ZERO, zeroData)
+
+      const sibs: Buffer[] = [zeroHash]
+      sibs.push(
+        hashFunction(
+          hashBuffer.fill(hashFunction(zeroData), 0, 32).fill(zeroHash, 32)
+        )
+      )
+      assert(
+        await tree.verifyAndStore({
+          rootHash: await tree.getRootHash(),
+          value: SparseMerkleTreeImpl.emptyBuffer,
+          key: TWO,
+          siblings: sibs.reverse(),
+        }),
+        'Verifying and storing empty data at TWO failed'
+      )
+
+      const twoData: Buffer = Buffer.from('TWO 2')
+      assert(await tree.update(TWO, twoData), 'Updating data at TWO failed')
+
+      // Check Proof for ZERO
+      let proof: MerkleTreeInclusionProof = await tree.getMerkleProof(
+        ZERO,
+        zeroData
+      )
+
+      assert(proof.value.equals(zeroData))
+      assert(proof.key.equals(ZERO))
+
+      assert(proof.siblings.length === 2)
+
+      let hash: Buffer = zeroHash
+      assert(proof.siblings[1].equals(hash))
+      hash = hashFunction(
+        hashBuffer.fill(hashFunction(twoData), 0, 32).fill(zeroHash, 32)
+      )
+      assert(proof.siblings[0].equals(hash))
+      assert(proof.rootHash.equals(await tree.getRootHash()))
+
+      // Check Proof for TWO
+      proof = await tree.getMerkleProof(TWO, twoData)
+
+      assert(proof.value.equals(twoData))
+      assert(proof.key.equals(TWO))
+
+      assert(proof.siblings.length === 2)
+
+      hash = zeroHash
+      assert(proof.siblings[1].equals(hash))
+      hash = hashFunction(
+        hashBuffer.fill(hashFunction(zeroData), 0, 32).fill(zeroHash, 32)
+      )
+      assert(proof.siblings[0].equals(hash))
+
+      assert(proof.rootHash.equals(await tree.getRootHash()))
+    })
+  })
```

Comment:
Went through the tests and they look good! I did write a little benchmark test in addition here:

```
    it.only('can make a whole lot of updates at random places in the tree', async () => {
      const tree: SparseMerkleTree = new SparseMerkleTreeImpl(
        db,
        undefined,
        24,
        hashFunction
      )

      const startTime = +new Date()
      for (let i = 0; i < 100; i++) {
        const key = new BigNumber(Math.floor(Math.random()*50000))
        await tree.update(key, Buffer.from('yo what is gucci'))
      }
      const finishTime = +new Date()
      const durationInMiliseconds = finishTime - startTime
      console.log('Duration:', durationInMiliseconds)
    }).timeout(9000)
```

Could consider adding that at the bottom but commented out, or just not at all because this benchmark is truly a hasty and ugly construction lol :P

Reply:
I added them and have them uncommented. They go very quickly, so I'm not too worried about having them in.